### PR TITLE
first draft of a date range filter for bones

### PR DESCRIPTION
**Pull request by skoegl**
Here I would like to introduce a date range filter for dateBones.

SInce only one kind entry is allowed to use $lt or $gt operators in filter, we should also have mutual exclusive filter groups. Here we go

### CODE REVIEW

File: bones/date.py [hunk 1]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
 	return skelStucture[boneName]["type"] == "date"
 
-#Register this Bone in the global queue
+
+class DateRangeFilterPlugin(html5.Div):
+	def __init__(self, extension, view, modul, *args, **kwargs):
+		super(DateRangeFilterPlugin, self).__init__(*args, **kwargs)
+		self.view = view
+		self.extension = extension
+		self.modul = modul
```
Comment (phorward) [inline bones/date.py]: please rename "modul" into "module"

File: bones/date.py [hunk 2]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+		title = html5.H2()
+		title.appendChild(html5.TextNode(extension["name"]))
+		self.appendChild(title)
+
+		# from daterange value
+		self.fromDatepicker = html5.Input()
+		self.fromDatepicker["type"] = "date"
+		self.fromDatepicker["id"] = "from-date-%s" % self.extension["target"]
+		self.fromDatepicker["class"] = "extended-date-range-filter"
+		fromLabel = html5.Label("Startdatum")
```
Comment (phorward) [inline bones/date.py]: Please provide `translate("begin date")` or something similar and put appropriate translations into translations/de.py

File: bones/date.py [hunk 3]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+		span.appendChild(fromLabel)
+		span.appendChild(self.fromDatepicker)
+		self.appendChild(span)
+
+		# to daterange value
+		self.toDatepicker = html5.Input()
+		self.toDatepicker["type"] = "date"
+		self.toDatepicker["id"] = "to-date-%s" % self.extension["target"]
+		self.toDatepicker["class"] = "extended-date-range-filter %s" % self.extension["target"]
+		toLabel = html5.Label("Enddatum")
```
Comment (phorward) [inline bones/date.py]: same here (translate)

File: bones/date.py [hunk 4]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+		toLabel["for"] = "to-date"
+		span2 = html5.Span()
+		span2.appendChild(toLabel)
+		span2.appendChild(self.toDatepicker)
+		self.appendChild(span2)
+		self.clearButton = html5.ext.Button("Filter entfernen", self.clearFilter)
```
Comment (phorward) [inline bones/date.py]: translation

File: bones/date.py [hunk 5]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+		self.sinkEvent("onChange")
+
+	def clearFilter(self, fire=True):
+		self.fromDatepicker["value"] = None
+		self.toDatepicker["value"] = None
+		if fire:
+			self.filterChangedEvent.fire()
+
+	def updateFilter(self, filter, mutualExclusiveGroupTarget=None, mutualExclusiveGroupKey=None):
+		print("updateFilter", self.extension["target"], filter, mutualExclusiveGroupTarget)
```
Comment (phorward) [inline bones/date.py]: please remove / convert into comment

File: bones/date.py [hunk 6]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+			self.clearFilter(False)
+			return filter
+
+		clearedFilter = {}
+		for key, value in filter.items():
+			if key.find("$") != -1 and key.find(self.extension["target"]) == -1:
+				continue
+			else:
+				clearedFilter[key] = value
+		print("clearedFilter before new values", self.extension["target"], clearedFilter)
```
Comment (phorward) [inline bones/date.py]: comment out

File: bones/date.py [hunk 7]
```diff
@@ -202,10 +205,94 @@ def serializeForPost(self):
...
+		if self.fromDatepicker["value"] and self.toDatepicker["value"]:
+			clearedFilter["orderby"] = self.extension["target"]
+			clearedFilter["%s$gt" % self.extension["target"]] = self.fromDatepicker["value"]
+			clearedFilter["%s$lt" % self.extension["target"]] = self.toDatepicker["value"]
+		print("clearedFilter result", self.extension["target"], clearedFilter)
```
Comment (phorward) [inline bones/date.py]: comment out

**Review by phorward — CHANGES_REQUESTED** [review]
Hi @skoegl! 
This looks good to me, can you please change the commented lines and add translations?

**Comment by skoegl** [comment]
I'm myself not completely satisfied with this PR. We should discuss especially the mutual exclusive group thing in our next viur meeting or here. Alternatively it works fine for now, so let's accept to get this feature into develop and discuss improvement in a next PR.

**Comment by phorward** [comment]
> Alternatively it works fine for now, so let's accept to get this feature into develop and discuss improvement in a next PR.

That's a good deal. I'm also of the opinion that new features should be made available ASAP... improvements and perfection can be done later. I will take a look at it how it's generally running in the next few days and then merge it into dev.

**Review by phorward — APPROVED** [review]

**Pull request merged**